### PR TITLE
Restrict undefined grids from UPP computation Part 3

### LINKED ISSUE #347: Runtime error when running a regression test in debug mode
Please see the related issue https://github.com/ufs-community/ufs-weather-model/issues/686

### PR DESCRIPTION
**Pull request by WenMeng-NOAA**
The main changes are:

- Modify boundary layer computation BNDLYR.f which causes ufs-weather-model global inline post debug mode failure.
- Enhancement of undefined grids involved UPP computation which cause regional inline post debug mode failure.
- Remove the legacy sigma level temperature from RRFS products. 
- Solve the extreme values in MCONV issue of RRFS products with output domain larger than computation domain. 
- Increase VERSION to 10.0.9.

### CODE REVIEW

File: sorc/ncep_post.fd/ALLOCATE_ALL.f [hunk 1]
```diff
@@ -46,7 +46,7 @@ SUBROUTINE ALLOCATE_ALL()
       integer ierr,jsx,jex
       integer i,j,l,k
 ! Allocate arrays
-      allocate(u(im,jsta_2l:jend_2u,lm))
+      allocate(u(im+1,jsta_2l:jend_2u,lm))
```
Comment (junwang-noaa) [inline sorc/ncep_post.fd/ALLOCATE_ALL.f:49]: Is u index im+1 used in the code? I am curious why only u field has this dimension.

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/ALLOCATE_ALL.f:49]: @junwang-noaa The arrays U/V are used by rap/hrrr for their special grids. You may see the read interface  INITPOST.F for rap/hrrr. In my last PR for initializing all post internal variables, I accidentally change it.

**Review by EricRogers-NOAA — APPROVED** [review]

File: sorc/ncep_post.fd/SLP_new.f [hunk 2]
```diff
@@ -258,7 +259,12 @@ SUBROUTINE MEMSLP(TPRES,QPRES,FIPRES)
         DO J=JSTA_M,JEND_M
           DO I=2,IM-1
 
-           if(PSLP(I,J)<spval) then
+           !if(PSLP(I,J)<spval) then
+           if(HTM2D(I-1,J)<spval.and.HTM2D(I+1,J)<spval.and. &
+              HTM2D(I,J-1)<spval.and.HTM2D(I,J+1)<spval.and. &
+              HTM2D(I-1,J-1)<spval.and.HTM2D(I+1,J-1)<spval.and. &
+              HTM2D(I-1,J+1)<spval.and.HTM2D(I+1,J+1)<spval.and. &
+              PSLP(I,J)<spval)then
```
Comment (HuiyaChuang-NOAA) [inline sorc/ncep_post.fd/SLP_new.f]: I think you should use QPRES and TPRES instead of PSLP

File: sorc/ncep_post.fd/ALLOCATE_ALL.f (hunk 1 above)
Comment (HuiyaChuang-NOAA) [reply sorc/ncep_post.fd/ALLOCATE_ALL.f:49]: UH is used for other models and it does have im dimensiion

File: sorc/ncep_post.fd/CALLCL.f [hunk 3]
```diff
@@ -88,6 +89,7 @@ SUBROUTINE CALLCL(P1D,T1D,Q1D,PLCL,ZLCL)
       DO 30 J=JSTA_M,JEND_M
       DO 30 I=2,IM-1
 !     DO 30 I=1,IM
+      IF(P1D(I,J)<spval.and.Q1D(I,J)<spval.and.Q1D(I,J)<spval)THEN
```
Comment (HuiyaChuang-NOAA) [inline sorc/ncep_post.fd/CALLCL.f]: you have two "Q1D" here :-)

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/CALLCL.f]: @HuiyaChuang-NOAA Thanks for catching this. Will update.

File: sorc/ncep_post.fd/SLP_new.f (hunk 2 above)
Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/SLP_new.f]: @HuiyaChuang-NOAA Thanks for comments. Will update it.

File: sorc/ncep_post.fd/CLDRAD.f [hunk 4]
```diff
@@ -2732,10 +2732,13 @@ SUBROUTINE CLDRAD
 !		  qqi(i,j,k)=qqw(i,j,k) ! because GFS only uses cloud water
 !		  qqw(i,j,k)=0.
 !		 end if 
+                 if(pint(i,j,k)<spval.and.qqw(i,j,k)<spval.and. &
```
Comment (HuiyaChuang-NOAA) [inline sorc/ncep_post.fd/CLDRAD.f:2735]: please add pint(i,j,k+1)

File: sorc/ncep_post.fd/FRZLVL2.f [hunk 5]
```diff
@@ -238,6 +240,11 @@ SUBROUTINE FRZLVL2(ISOTHERM,ZFRZ,RHFRZ,PFRZL)
                RHFRZ(I,J) = AMAX1(0.01,RHFRZ(I,J))
                RHFRZ(I,J) = AMIN1(RHFRZ(I,J),1.00)
                ZFRZ(I,J)  = AMAX1(0.0,ZFRZ(I,J))
+         ELSE
+               RHFRZ(I,J) = spval
+               RHFRZ(I,J) = spval
```
Comment (HuiyaChuang-NOAA) [inline sorc/ncep_post.fd/FRZLVL2.f]: can remove this line as it's the same as the line above

File: sorc/ncep_post.fd/MIXLEN.f [hunk 6]
```diff
@@ -124,6 +136,7 @@ SUBROUTINE MIXLEN(EL0,EL)
 !$omp  parallel do private(i,j,elst,elvgd,ensq,q2kl,ziag)
         DO J=JSTA_M,JEND_M
           DO I=2,IM-1
+            IF(T(I,J,L)<spval)THEN
```
Comment (HuiyaChuang-NOAA) [inline sorc/ncep_post.fd/MIXLEN.f:139]: I suggest you add PMID

File: sorc/ncep_post.fd/WETBULB.f [hunk 7]
```diff
@@ -59,13 +60,14 @@ SUBROUTINE WETBULB(T,Q,PMID,HTM,KARR,TWET)
...
 !-----------------------------------------------------------------------
-                             DO 300 L=1,LM
+      DO 300 L=1,LM
       DO 125 J=JSTA,JEND
       DO 125 I=1,IM
         IF (HTM(I,J,L)<1.0) THEN
           THESP(I,J)=273.15
           cycle    
         ENDIF
+        IF(T(I,J,L)<spval)THEN
```
Comment (HuiyaChuang-NOAA) [inline sorc/ncep_post.fd/WETBULB.f:70]: suggest add q and pmid

**Review by HuiyaChuang-NOAA — COMMENTED** [review]
Made a few comments, let me know what you think

File: sorc/ncep_post.fd/FRZLVL2.f (hunk 5 above)
Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/FRZLVL2.f]: @HuiyaChuang-NOAA I should mean ZFRZ.

File: sorc/ncep_post.fd/WETBULB.f (hunk 7 above)
Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/WETBULB.f:70]: @HuiyaChuang-NOAA Basically for these areas not in model computation domain, if T is undefined, Q, P, U/V will be the same values. But adding more constraints would be safe. Thanks!

**Review by HuiyaChuang-NOAA — APPROVED** [review]

**Comment by WenMeng-NOAA** [comment]
The ufs-weather-model tests for global inline post, global inline post in debug mode, regional inline post and regional inline post in debug mode were conducted on WCOSS-Dell and Hera. All tests can be successfully run though.

The UPP regression tests were conducted on  WCOSS-Dell, WCOSS-Cray and Hera. With the PR merged, the baselines for 
hafs, fv3r will be recreated.

**Comment by HuiyaChuang-NOAA** [comment]
> The ufs-weather-model tests for global inline post, global inline post in debug mode, regional inline post and regional inline post in debug mode were conducted on WCOSS-Dell and Hera. All tests can be successfully run though.
> 
> The UPP regression tests were conducted on WCOSS-Dell, WCOSS-Cray and Hera. With the PR merged, the baselines for
> hafs, fv3r will be recreated.

Great news!  Thank you for your effort.

**Pull request merged**